linode/manager
# E2E Test Setup & Teardown Refactor

* This is a significant refactor that changes the design pattern used for creating e2e tests. The idea is that now, all setup/teardown code should be baked into a test spec file. 

* To combat unexpected failures resulting in an unexpected account state after a test run, the setup/teardown utils use the v4 API to create / delete things like linodes, domains, volumes, etc. 
* Also used this PR as an opportunity to fix issues plaguing the docker e2e. We now call `yarn build && yarn serve` to first build the static assets then start a simple https server to run the manager app. This prevents docker container start timing issues.


## Setup/Teardown Related Changes:

* Updated `.env.example` to warn contributors about running the e2e tests
* Added custom command for `readToken` to get the token value from `./localStorage.json`
* Added custom commands that call api create/delete utilities. Needed because any async code used by `wdio` test runner has to be wrapped in a custom command like so: `browser.addCommand('someAsyncCommand', function async() {`
* Fixed issues with existing setup/teardown method's error logging 
* Added `apiCreateLinode` , `apiDeleteAllLinodes` and `apiDeleteAllVolumes` utilities
* Added command line flags to wdio that will set the baseUrl.. `--dev`, `--test`, `--staging`, `--prod`
* Removed manual "create linode" step of `setup.spec.js` script.. This script's purpose is to now only remove existing account data. 
* Updated various tests to use api create/delete utilities. 

## Docker Related Changes:
* Added `serve` dependency. one can now run `yarn build` followed by `yarn serve` to start manager and serve static files (Needed to fix timing issues on docker).
* Added docker entrypoint script for manager
* Updated Dockerfile image to latests node/debian version.

## Misc Fixes
* Fixed clone linode tests (no longer has set target handling)
* Fixed search bar tests

# To Test
Main thing we want here is to prove that creating/deleting linodes/volumes works in our tests

0. Make sure your `.env` file is setup using your dev test account credentials
1. `yarn && yarn start`
2. `yarn selenium` (in a new shell)
3. `yarn e2e` (in a new shell)

* I ran the tests yesterdayt and got the following results:
https://gist.github.com/patthiel/51724c061d431a5266cf443d4a272b7b

111 passing (1173.50s)
11 skipped
36 failing

I believe in this test run there were some timeouts which resulted in whole spec files failing all tests..

I ran the tests again this morning and got different results:

133 passing (1026.90s)
11 skipped
14 failing

https://gist.github.com/patthiel/3602604570f793b6e0e8255b26b722c3

* Some of these failures appear to be due to outdated tests or `data-qa` attributes that have been removed. I will address these in a followup PR, as this one is already huge. 

# To Test the Static Server:

```bash
yarn # to install serve
yarn build # build dist files
yarn serve # should start a https server (https is needed for docker tests)
```
